File: src/day6.rs
use std::fs;
use std::collections::{HashMap, HashSet};

fn part1(lines: &[&str]) -> i32 {
    let mut groups: Vec<i32> = Vec::new();
    let mut group = HashSet::new();
    for line in lines {
        if *line == "" {
            groups.push(group.len() as i32);
            group = HashSet::new();
        }
        else {
            for c in line.chars() {
                group.insert(c);
            }
        }
    }
    return groups.iter().sum();
}

fn part2(lines: &[&str]) -> i32 {
    let mut groups: Vec<i32> = Vec::new();
    let mut group = HashMap::new();
    let mut count = 0;
    for line in lines {
        if *line == "" {
            let mut n = 0;
            for (_k,v) in group {
                if v == count {
                    n += 1;
                }
            }
            groups.push(n);
            group = HashMap::new();
            count = 0
        }
        else {
            for c in line.chars() {
                let x = group.entry(c).or_insert(0);
                *x += 1;
            }
            count += 1;
        }
    }
    return groups.iter().sum();
}

pub fn day6() {
    let contents = fs::read_to_string("day6.txt").unwrap();
    let input: Vec<&str> = contents.lines().collect();
    let mut out = part1(&input);
    println!("{}", out);
    assert_eq!(out, 6714);
    out = part2(&input);
    println!("{}", out);
    assert_eq!(out, 3435);
}

File: 2020/day6.py
#!/usr/bin/env python3

def part1(lines):
    groups = []
    group = set()
    for line in lines:
        line = line.strip()
        if line == "":
            groups.append(len(group))
            group = set()
        else:
            for l in line:
                group.add(l)

    return sum(groups)

def part2(lines):
    groups = []
    group = {}
    count = 0
    for line in lines:
        line = line.strip()
        if line == "":
            n = 0
            for k,v in group.items():
                if v == count:
                    n += 1
            groups.append(n)
            group = {}
            count = 0
        else:
            for l in line:
                if l in group:
                    group[l] += 1
                else:
                    group[l] = 1
            count += 1
    return sum(groups)

if __name__ == "__main__":
    with open("day6.txt") as f:
        lines = [line.strip() for line in f.readlines()]

    sample = """abc

a
b
c

ab
ac

a
a
a
a

b

"""
    assert part1(sample.split('\n')) == 11
    print(part1(lines))

    assert part2(sample.split('\n')) == 6
    print(part2(lines))
